core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -593,9 +722,147 @@ private Builder visitAggregate(Aggregate e, List<Integer> groupKeyList,
+  }
+
+  // Function to extract column names used from the subQuery in the GROUP BY clause
+  private List<String> extractColumnNamesUsedFromSubQuery(List<SqlNode> groupByList, String subQueryAlias) {
+    List<String> columnNamesUsedFromSubQuery = new ArrayList<>();
+    for (SqlNode groupByNode : groupByList) {
+      if (groupByNode instanceof SqlIdentifier) {
+        if (((SqlIdentifier) groupByNode).names.get(0).equals(subQueryAlias)) {
+          columnNamesUsedFromSubQuery.add(((SqlIdentifier) groupByNode).names.get(1));
+        }
+      }
+    }
+    return columnNamesUsedFromSubQuery;
+  }
+
+  // Function to create a list of SqlIdentifier for the identified column names
+  private List<SqlIdentifier> createSqlIdentifiers(List<String> columnNamesUsedFromSubQuery) {
+    List<SqlIdentifier> sqlIdentifierList = new ArrayList<>();
+    for (String col : columnNamesUsedFromSubQuery) {
+      sqlIdentifierList.add(new SqlIdentifier(ImmutableList.of(col), POS));
+    }
+    return sqlIdentifierList;
+  }
+
+  // Function to update the SELECT list in the subQuery
+  private void updateSelectListInSubQuery(
+      SqlBasicCall sqlBasicCall, List<SqlIdentifier> sqlIdentifiersNew) {
+    List<String> columnNamesInSelect = new ArrayList<>();
+    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
+      if (sqlNode instanceof SqlSelect) {
+        SqlNodeList sqlNodeList = ((SqlSelect) sqlNode).getSelectList();
+        if (sqlNodeList != null) {
+          for (SqlNode selectNode : sqlNodeList) {
+            if (selectNode instanceof SqlIdentifier) {
+              if (((SqlIdentifier) selectNode).names.size() > 1) {
+                columnNamesInSelect.add(((SqlIdentifier) selectNode).names.get(1));
+              } else {
+                columnNamesInSelect.add(((SqlIdentifier) selectNode).names.get(0));
+              }
+            }
+          }
+          if (sqlIdentifiersNew.size() > sqlNodeList.size()) {
+            for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
+              if (!columnNamesInSelect.contains(sqlIdentifier.names.get(0))) {
+                sqlNodeList.add(createNewSelectItem(sqlIdentifier, sqlNode));
+              }
+            }
+            ((SqlSelect) sqlNode).setSelectList(new SqlNodeList(sqlNodeList, POS));
+          }
+        } else if(sqlIdentifiersNew.size() > 0) {
+          SqlNodeList sqlNodes = new SqlNodeList(POS);
+          for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
+            if((((SqlSelect) sqlNode).getFrom() instanceof SqlJoin)) {
+              sqlNodes.add(createNewSelectItem(sqlIdentifier, sqlNode));
+              ((SqlSelect) sqlNode).setSelectList(new SqlNodeList(sqlNodes, POS));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Function to create a new SqlSelectItem with an alias if it ends with a digit
+  private SqlNode createNewSelectItem(SqlIdentifier sqlIdentifier, SqlNode sqlNode) {
+    if (endsWithDigit(sqlIdentifier.names.get(0))) {
```

Comment:
Please refactor this code 
'private SqlNode createNewSelectItem(SqlIdentifier sqlIdentifier, SqlNode sqlNode) {
    ImmutableList<String> si = getOperandNames(sqlNode);

    if (endsWithDigit(sqlIdentifier.names.get(0))) {
        return createAsSqlNode(si.get(0), sqlIdentifier.names.get(0));
    } else {
        return new SqlIdentifier(ImmutableList.of(si.get(0), sqlIdentifier.names.get(0)), SqlParserPos.ZERO);
    }
}

private ImmutableList<String> getOperandNames(SqlNode sqlNode) {
    SqlNode[] operands = getOperandsFromJoin(sqlNode);
    return ((SqlIdentifier) operands[1]).names;
}

private SqlNode[] getOperandsFromJoin(SqlNode sqlNode) {
    if (sqlNode instanceof SqlSelect) {
        SqlJoin joinNode = (SqlJoin) ((SqlSelect) sqlNode).getFrom();
        SqlBasicCall basicCall = (SqlBasicCall) (joinNode.getRight() != null ? joinNode.getRight() : joinNode.getLeft());
        return basicCall.getOperands();
    }
    return null;
}

private SqlNode createAsSqlNode(String alias, String columnName) {
    return SqlStdOperatorTable.AS.createCall(POS,
            ImmutableList.of(new SqlIdentifier(ImmutableList.of(alias, columnName.replaceAll("\\d+$", "")), SqlParserPos.ZERO),
                    new SqlIdentifier(columnName, SqlParserPos.ZERO)));
}
'

Reply:
Done